allennlp_semparse/state_machines/states/grammar_statelet.py

```diff
@@ -128,7 +128,12 @@ def __eq__(self, other):
                 [
                     self._nonterminal_stack == other._nonterminal_stack,
                     util.tensors_equal(self._valid_actions, other._valid_actions),
-                    self._is_nonterminal == other._is_nonterminal,
+                    all(
+                        [
+                            self._is_nonterminal.__module__ == other._is_nonterminal.__module__,
+                            self._is_nonterminal.__name__ == other._is_nonterminal.__name__,
+                        ]
+                    ),
```

Comment:
Not sure if there's a better way to do this, but without this change, tests will fail against master branch of allennlp. Specifically `test_model_can_train_save_and_load` tests.

Reply:
Can you point me to the error?  I'm not sure why this would be necessary, unless it's a python version issue, where with newer versions of python the old comparison doesn't work.  It looked like the only thing failing in your initial commit was black.

Reply:
The error comes from here:

https://github.com/allenai/allennlp/blob/90f003799e9fd6f786f54027441600ac6e871d20/allennlp/common/testing/model_test_case.py#L171-L173

The error arises because `self._is_nonterminal == other._is_nonterminal` returns `False`. Could be a difference between Python 3.7 and 3.8, I'm not sure yet.

Reply:
But if I look [here](https://github.com/allenai/allennlp-semparse/runs/1217061145), it looks like all of the tests pass.  Where are you seeing a failure?

Reply:
I first saw the error while I was running tests locally in Python 3.8 environment. I just updated the Python version in GH Actions to 3.8, and you can now see the errors. So it's definitely due to differences in Python versions.